google/cloud/spanner/samples/pg_samples.cc

```diff
@@ -0,0 +1,371 @@
+        auto update = update_budget(txn, 1, 1, *budget1 - transfer_amount);
+        if (!update) return std::move(update).status();
+        update = update_budget(txn, 2, 2, *budget2 + transfer_amount);
+        if (!update) return std::move(update).status();
+        return google::cloud::spanner::Mutations{};
+      });
+  if (!commit) throw std::runtime_error(commit.status().message());
+  std::cout << "Update was successful.\n";
+}
+// [END spanner_postgresql_dml_getting_started_update]
+
+void DropDatabase(google::cloud::spanner_admin::DatabaseAdminClient client,
+                  google::cloud::spanner::Database const& database) {
+  auto status = client.DropDatabase(database.FullName());
+  if (!status.ok()) throw std::runtime_error(status.message());
+  std::cout << "Database " << database << " dropped.\n";
+}
+
+}  // namespace samples
+
+google::cloud::spanner::Database Database(std::vector<std::string> argv) {
+  if (argv.size() != 4) {
+    std::string args = " <project-id> <instance-id> <database-id>";
+    throw std::runtime_error(argv[0] + args);
+  }
+  google::cloud::spanner::Instance instance(std::move(argv[1]),
+                                            std::move(argv[2]));
+  return google::cloud::spanner::Database(instance, std::move(argv[3]));
+}
+
+using CommandType = std::function<void(std::vector<std::string>)>;
+
+CommandType Command(
+    void (*sample)(google::cloud::spanner_admin::DatabaseAdminClient,
+                   google::cloud::spanner::Database const&)) {
+  return [sample](std::vector<std::string> argv) {
+    auto database = Database(std::move(argv));
+    google::cloud::spanner_admin::DatabaseAdminClient client(
+        google::cloud::spanner_admin::MakeDatabaseAdminConnection());
+    sample(std::move(client), database);
+  };
+}
+
+CommandType Command(void (*sample)(google::cloud::spanner::Client)) {
+  return [sample](std::vector<std::string> argv) {
+    auto database = Database(std::move(argv));
+    google::cloud::spanner::Client client(
+        google::cloud::spanner::MakeConnection(database));
+    sample(std::move(client));
+  };
+}
+
+int RunOneCommand(std::vector<std::string> argv) {
+  std::map<std::string, CommandType> commands = {
+      {"create-database", Command(samples::CreateDatabase)},
+      {"create-tables", Command(samples::CreateTables)},
+      {"add-column", Command(samples::AddColumn)},
+      {"create-storing-index", Command(samples::CreateStoringIndex)},
+      {"insert-data", Command(samples::InsertData)},
+      {"query-with-parameter", Command(samples::QueryWithParameter)},
+      {"dml-getting-started-update", Command(samples::DmlGettingStartedUpdate)},
+      {"drop-database", Command(samples::DropDatabase)},
+  };
+  auto it = commands.find(argv[0]);
+  if (it == commands.end()) {
+    throw std::runtime_error(argv[0] + ": Unknown command");
+  }
+  it->second(std::move(argv));
+  return 0;
+}
+
+void SampleBanner(std::string const& name) {
```

Comment:
We *could* factor this out as it appears in `samples.cc` as well. (We could also refactor the `Command`/`make_command_entry` stuff too). Probably not something we want to do in this PR. Maybe not something we ever want to do.

Reply:
I mentioned possible refactorings down the line in the original PR description, but, yes, we should certainly postpone that decision until we see where things fall.  (At the moment it looks like it would be a net complication.)